src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs

```diff
@@ -262,4 +262,47 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
         public override int GetHashCode() =>
             GetType().GetHashCode();
     }
+
+    // This class exists to be EqualityComparer<string>.Default. It can't just use the GenericEqualityComparer<string>,
+    // as it needs to also implement IAlternateEqualityComparer<ReadOnlySpan<char>, string>, and it can't be
+    // StringComparer.Ordinal, as that doesn't derive from the abstract EqualityComparer<T>.
+    [Serializable]
+    internal sealed partial class StringEqualityComparer :
+        EqualityComparer<string>,
+        IAlternateEqualityComparer<ReadOnlySpan<char>, string>,
+        ISerializable
+    {
+        void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context)
+        {
+            // This type was introduced after BinaryFormatter was deprecated. BinaryFormatter serializes
+            // types from System.Private.CoreLib without an assembly name, and then on deserialization assumes
+            // "mscorlib". To support such roundtripping, this type would need to be public and type-forwarded from
```

Comment:
Nit: It would not be backward compatible. The BF expectations have been that you can serialize on version N and deserialize on version N-1. I think this is no different from other places that use `info.SetType` in this file for backward compatibility, like https://github.com/dotnet/runtime/blob/9a4329d65c4d272e26b2c693341ef4b26a5bc8c8/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs#L245-L249 .

Reply:
Thanks but I'm not following. This isn't for back compat, but rather to be able to still serialize instances produced by `EqualityComparer<string>.Default` and in particular collections that contain one. To make it fully roundtrippable back to the same type, the only way I know is is to make it public and expose it in the mscorelib shim, and that doesn't seem worthwhile. Are you disagreeing, or is there a better way, or the not is with something in the comment? Thanks.

Reply:
The two relevant scenarios that are expected to work with BF are:

1. Serialize `new Dictionary<string>()` on .NET 9 and deserialize it on .NET 8 or even on .NET Framework
2. Serialize `new Dictionary<string>()` on .NET 9 and deserialize it on .NET 9+ with the new StringComparer roundtripped

In the current state of the PR, 1 works and 2 is broken. If you were to go with the alternative to plumb this through mscorlib, 1 is going to be broken and 2 is going to work. I do not think there is a way to make both 1. and 2. work with the BF transparently. It would require a new BF feature. The current state of the PR (1 works and 2 is broken) is a fine choice given that BF is deprecated.

The comment in the code makes it sound like that doing the extra work to plumb this through mscorlib would be an improvement. I was trying to point out that it would make one scenario work and broke the other scenario.

Reply:
Ah, ok, thanks for clarifying. I'll fix the comment.